File: app/src/main/java/tech/thdev/mediaprojectionexample/MainActivity.java
package tech.thdev.mediaprojectionexample;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import butterknife.Bind;
import butterknife.OnClick;
import tech.thdev.mediaprojectionexample.base.view.BaseActivity;
import tech.thdev.mediaprojectionexample.service.VideoViewService;

/**
 * MediaProjection Sample main
 * <p/>
 * Created by Tae-hwan on 4/8/16.
 */
public class MainActivity extends BaseActivity {

    public static final int REQ_CODE_OVERLAY_PERMISSION = 9999;

    @Bind(R.id.fab)
    FloatingActionButton floatingActionButton;

    @Override
    protected int getContentLayoutResource() {
        return R.layout.activity_main;
    }

    @Override
    protected void onCreate() {
        floatingActionButton = (FloatingActionButton) findViewById(R.id.fab);
        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @OnClick(R.id.btn_start_media_projection_service)
    public void onClickStartMediaProjection(View view) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
            // User rights acquired
            Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + getPackageName()));
            startActivityForResult(intent, REQ_CODE_OVERLAY_PERMISSION);

        } else {
            startVideoViewService();
        }
    }

    @OnClick(R.id.btn_start_media_projection_activity)
    public void onClickStartMediaProjectionActivity(View view) {
        startActivity(new Intent(this, MediaProjectionActivity.class));
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQ_CODE_OVERLAY_PERMISSION:
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    if (Settings.canDrawOverlays(this)) {
                        startVideoViewService();
                    }
                }
                break;

            default:
                break;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private void startVideoViewService() {
        startService(new Intent(this, VideoViewService.class));
    }
}

File: app/src/main/java/tech/thdev/mediaprojectionexample/ui/main/MainActivity.kt
package tech.thdev.mediaprojectionexample.ui.main

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import androidx.activity.result.contract.ActivityResultContract
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import tech.thdev.mediaprojectionexample.R
import tech.thdev.mediaprojectionexample.databinding.ActivityMainBinding
import tech.thdev.mediaprojectionexample.databinding.ContentMainBinding
import tech.thdev.mediaprojectionexample.ui.mediaprojection.MediaProjectionActivity
import tech.thdev.mediaprojectionexample.ui.service.VideoViewService

/**
 * MediaProjection Sample main
 *
 * Created by Tae-hwan on 4/8/16.
 */
class MainActivity : AppCompatActivity() {

    private val canOverlay = registerForActivityResult(OverlayActivityResultContract()) {
        runService()
    }

    private fun runService() {
        if (Settings.canDrawOverlays(this)) {
            startService()
        } else {
            showRejectDialog()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityMainBinding.inflate(layoutInflater)
        val contentMainBinding = ContentMainBinding.inflate(layoutInflater, binding.root, true)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)

        contentMainBinding.btnStartMediaProjectionService.setOnClickListener {
            runService()
        }

        contentMainBinding.btnStartMediaProjectionActivity.setOnClickListener {
            startActivity(MediaProjectionActivity.newInstance(this))
        }
    }

    private fun showRejectDialog() {
        AlertDialog.Builder(this)
            .setTitle(R.string.reject_dialog_title)
            .setMessage(R.string.reject_dialog_message)
            .setPositiveButton("Ok") { _, _ ->
                canOverlay.launch("package:$packageName")
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun startService() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(VideoViewService.newService(this))
        } else {
            startService(VideoViewService.newService(this))
        }
    }
}

class OverlayActivityResultContract : ActivityResultContract<String, Boolean>() {

    override fun createIntent(context: Context, input: String?): Intent =
        Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse(input))

    override fun parseResult(resultCode: Int, intent: Intent?): Boolean =
        resultCode == Activity.RESULT_OK
}